hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/throttle/CompactionThroughputControllerFactory.java

```diff
@@ -47,24 +53,35 @@ private CompactionThroughputControllerFactory() {
 
   public static ThroughputController create(RegionServerServices server,
       Configuration conf) {
-    Class<? extends ThroughputController> clazz = getThroughputControllerClass(conf);
-    ThroughputController controller = ReflectionUtils.newInstance(clazz, conf);
-    controller.setup(server);
+    Class<? extends ThroughputController> clazz =
+        getThroughputControllerClass(conf);
+    ThroughputController controller = controllerMap.get(clazz);
+    if (controller == null) {
+      controller = ReflectionUtils.newInstance(clazz, conf);
+      controller.setup(server);
+      controllerMap.put(clazz, controller);
+    } else {
```

Comment:
why change it into this?

Reply:
> why change it into this?

if the situation is as follows:
1.we start RS with PressureAwareCompactionThroughputController
2.then change the conf to NoLimitThroughputController ,and execute 'update_config'
3.then change back the conf to PressureAwareCompactionThroughputController,and execute 'update_config'

In master branch now , step 2 and step 3 all create the new contoller object.
But the compaction started before step1 is still use the old Controller object(can not be update).

So I want to cache the controller to a map ,
so that we can execute 'update_config' to update the controller of the compaction started before step1。

step 1 and step 3 will use the same controller object in this patch.

Reply:
So this means we will never actually stop a controller?

Reply:
> So this means we will never actually stop a controller?

Because when we execute 'update_config', the running compaction is still use the old Controller object.
So I cache the controller in a map.keep there is always only one PressureAwareCompactionThroughputController object in RS.

If we stop controller after 'update_config' everytime,the running compaction still use the old controller object,and new compaction will use the new controller object.
the total higher.bound,lower.bound maybe more than we expected.

How do you think of it？